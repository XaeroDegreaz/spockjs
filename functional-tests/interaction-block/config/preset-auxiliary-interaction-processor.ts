import { InteractionProcessor } from '@spockjs/config';

export const assertionPostProcessors = [];

export const interactionProcessors: InteractionProcessor[] = [
  t => ({
    primary: false,
    declare: () => t.emptyStatement(),
    verify: () => t.emptyStatement(),
  }),
];
