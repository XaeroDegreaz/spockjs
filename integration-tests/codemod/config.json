{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert"
      }
    ]
  ]
}
