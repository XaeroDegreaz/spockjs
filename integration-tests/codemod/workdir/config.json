{
  "plugins": [
    [
      "@spockjs/babel-plugin-spock",
      {
        "powerAssert": false,
        "autoImport": "assert",
        "assertFunctionName": "assert",
        "presets": ["@spockjs/preset-sinon-mocks"]
      }
    ]
  ]
}
