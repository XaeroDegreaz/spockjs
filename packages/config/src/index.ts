import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

import { Hooks } from './hooks';

/**
 * Config as it can be provided by the user.
 * Comments indicate transformations made to conform to InternalConfig format.
 */
export const configType = t.exact(
  t.partial({
    powerAssert: t.boolean,
    autoImport: t.union([t.boolean, t.string]),
    staticTruthCheck: t.boolean,
    assertFunctionName: t.string,
    presets: t.array(t.string),
  }),
  'Config',
);
export type Config = t.TypeOf<typeof configType>;

/**
 * Config that disables all but the core set of features.
 */
export const minimalConfig: Config = {
  powerAssert: false,
  autoImport: false,
};

/**
 * Config containing the settings used if they are missing in the user-specified config.
 */
export const defaultConfig = {
  powerAssert: true,
  autoImport: true,
  staticTruthCheck: false,
  assertFunctionName: '',
  presets: [],
};

/**
 * Config after sanitizing in extractConfigFromState.
 */
export interface InternalConfig {
  readonly powerAssert: boolean;
  readonly autoImport: string; // empty => no auto import
  readonly staticTruthCheck: boolean;
  readonly assertFunctionName: string; // empty => no custom assert function name
  readonly hooks: Hooks;
}

export const extractConfigFromState = ({
  opts,
}: {
  opts: any;
}): InternalConfig => {
  const decoded = configType.decode(opts);

  let {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    presets,
  } = decoded.getOrElseL(errors => {
    throw new Error(
      ['Invalid plugin configuration:', ...failure(errors)].join('\n'),
    );
  });

  // powerAssert
  if (powerAssert === undefined) {
    ({ powerAssert } = defaultConfig);
  }

  // autoImport
  if (autoImport === undefined) {
    ({ autoImport } = defaultConfig);
  }
  if (autoImport === false) {
    autoImport = '';
  }
  if (autoImport === true) {
    autoImport = 'power-assert';
  }

  // staticTruthCheck
  if (staticTruthCheck === undefined) {
    ({ staticTruthCheck } = defaultConfig);
  }

  // assertFunctionName
  if (assertFunctionName === undefined) {
    ({ assertFunctionName } = defaultConfig);
  }

  // presets
  if (presets === undefined) {
    ({ presets } = defaultConfig);
  }
  const hooks = presets.map(preset => require(preset) as Hooks).reduce(
    (accHooks, moreHooks) => ({
      assertionPostProcessors: [
        ...accHooks.assertionPostProcessors,
        ...moreHooks.assertionPostProcessors,
      ],
    }),
    { assertionPostProcessors: [] },
  );

  return {
    powerAssert,
    autoImport,
    staticTruthCheck,
    assertFunctionName,
    hooks,
  };
};
