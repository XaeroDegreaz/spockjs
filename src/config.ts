export interface Config {
  powerAssert: boolean;
}

const defaultConfig: Config = {
  powerAssert: true,
};

export default defaultConfig;
