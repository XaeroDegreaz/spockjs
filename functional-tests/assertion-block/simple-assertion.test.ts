import { transform, TransformOptions } from '@babel/core';
import assert, { AssertionError } from 'assert';

import plugin from '@spockjs/babel-plugin-spock';
import { minimalConfig } from '@spockjs/config';

const babelOpts: TransformOptions = {
  plugins: [[plugin, minimalConfig]],
  highlightCode: false,
  filename: 'test.js',
};

test('assertifies an "expect"-labeled expression statement', () => {
  const { code } = transform(`expect: 1 === 2;`, babelOpts);
  expect(() => new Function('assert', code as string)(assert)).toThrow(
    AssertionError,
  );
});

test('assertifies a "then"-labeled expression statement', () => {
  const { code } = transform(`then: 1 === 2;`, babelOpts);
  expect(() => new Function('assert', code as string)(assert)).toThrow(
    AssertionError,
  );
});

test('does not assertify a "when"-labeled expression statement', () => {
  const { code } = transform(`when: 1 === 2;`, babelOpts);
  expect(() => new Function('assert', code as string)(assert)).not.toThrow();
});

test('assertifies all expression statements in a labeled block statement', () => {
  const { code } = transform(
    `expect: {
      1 === 1;
      0.5 * 4 < 3;
      ((x, y) => x === y)(42, 42);
    }`,
    babelOpts,
  );

  const customAssert = jest.fn(assert);
  new Function('assert', code as string)(customAssert);
  expect(customAssert).toHaveBeenCalledTimes(3);
});

test('assertifies inside an if / else statement', () => {
  const { code } = transform(
    `if(1 < 2) {
      expect: false;
    }
    else {
      expect: true;
    }`,
    babelOpts,
  );
  expect(() => new Function('assert', code as string)(assert)).toThrow(
    AssertionError,
  );
});

test('throws if a statement is not an expression statement', () => {
  expect(() => transform(`expect: if(true);`, babelOpts)).toThrowError(
    /Expected an expression statement, but got a statement of type IfStatement/,
  );
});
