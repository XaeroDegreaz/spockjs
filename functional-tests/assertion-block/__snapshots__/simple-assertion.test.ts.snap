// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if a statement is not an expression statement 1`] = `
"undefined: Expected an expression statement, but got a statement of type IfStatement
> 1 | expect: if(true);
    |         ^"
`;
